virtual_rainforest/core/config.py

```diff
@@ -432,7 +438,7 @@ def build_config(self) -> None:
         """
 
         # Get the config dictionaries
-        input_dicts = list(self.toml_contents.values())
+        input_dicts = list(self.toml_contents.values()) + self.extra_params
```

Comment:
Am I right in thinking that this step adds the extra parameters to the existing config dictionaries? If so is there anything here that handles the case where the config files define a parameter which is then supplied as an extra parameter? I'm not actually 100% sure what the desired behaviour would be in that case?

Reply:
If a parameter is set in one of the config files and you then *also* try to set it via a command-line arg, then you'll get an error, just as you would if you tried to set twice in different config files.

I can see that there might be a case where a user wants to override a parameter set in a config file, but I don't think that should be the default behaviour as it's a bit error prone. If we want to add this feature later we can do (e.g. we could have a separate `--param-override` flag for this case).

Reply:
Ahh yeah that makes sense! Also agree that overriding parameters seems error prone

Reply:
I'm not sure it is dangerous and I think there is a case for making the override behaviour the explicit mechanism for this approach. 

* The configuration process at the moment takes a bunch of files and uses those to build the configuration tree, substituting default values for any settings that aren't explicitly mentioned. That gives us a base configuration, with everything set. 

* If we want to tweak those settings, we shouldn't need to worry about whether the base files set a value or if it is just being populated by default. That would put us in a position of having to maintain multiple sets of base configuration files, each one omitting the setting that we want to tweak. It would be neater to have a default config that can just be adjusted as needed.

* Explicitly providing extra params would then be a way to substitute values in a particular run for a sensitivity analysis, by updating the `Config` object created from the base configuration for a given run. Those extra params aren't then simply added to the TOML contents, but are fed into a separate `Config` method. That does then need some kind of a validation step though.

Reply:
I think that method would be something like:

```python
    def override_config(self, extra_params) -> None:
        """Adjust an already created config to update settings

        Raises:
            ValidationError: if duplicate config definitions occur across files.
        """

        # Copy the current dictionary
        master = deepcopy(self.__dict__)
        
        # Update that dictionary
        # TODO - don't know if this works to replace values in master even when conflicts found, 
        #               or whether we need an explicit switch or method to do override.
        master, conflicts = config_merge(master, extra_params, conflicts=tuple())
        
        # Ignore any conflicts - explicitly overriding settings - so now revalidate against the 
        # schema defined in the base config.
        self.update(master)
        self.validate_config()
```

Reply:
At the moment, none of the constants are populated by default in the config (i.e. not using default values set within model schema). It's only after config validation when a set of constants is used to create a model specific constants class that default values come in, These defaults are hardcoded into the various constants classes, and just aren't replaced if there is no value stored in the config to replace them with.

So I don't really see a danger the current approach, unless we change how default constants are handled

Reply:
The `config_merge` does indeed return the updated dictionary when conflicts are present:

```python
In [11]: from virtual_rainforest.core.config import Config, config_merge
In [12]: cfg = Config('tests/core/data/all_config.toml')
In [13]: updated, conflicts = config_merge(c, {'core': {'layers': {'soil_layers': 3}}})
In [14]: updated['core']['layers']['soil_layers']
Out[14]: 3
```

So, this should work as:

```python
    def override_config(self, extra_params) -> None:
        """Adjust an already created config to update settings

        Raises:
            ValidationError: if duplicate config definitions occur across files.
        """
        
        # Update the dictionary
        updated, conflicts = config_merge(self, extra_params, conflicts=tuple())
        
        # Ignore any conflicts - explicitly overriding settings - so now revalidate against the 
        # schema defined in the base config.
        self.update(updated)
        self.validate_config()
```

Reply:
Obviously, if changing the grid size or another part of the config that does have a default setting through this method would be risky, but I don't think we would want to do that?

Reply:
So configuration settings that are then checked against other inputs/data could cause all sorts of chaos. 

But this process should only (?) ever be invoked when the `Config` is built and validated and bad overrides would cause exactly the same errors as bad initial settings - they can be JSONSchema invalid and get trapped early - or bad/inconsistent in setting up the rest of the simulation and cause a later error. 

Exposing a `Config.override_config` does raise a route where programmatic use _could_ change the `Config` half way through a simulation (and we should very clearly say that this is extremely ill-advised!), but in `vr_run` it would only ever be something like:

```python
config = Config(toml_files)
config.override_config(extra_params)
# Now leave config alone!
```